osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -120,44 +118,111 @@ private DbEncryptedSecret createEncrypted(final String deviceIdentification, fin
             throw new IllegalStateException("Could not decrypt secret (id: " + dbEncryptedSecret.getId() + ")", exc);
         }
     }
+
+    public void storeSecrets(final String deviceIdentification, final List<TypedSecret> secrets) {
+        //@formatter:off
+        secrets.stream()
+                .map(t -> this.validateNewSecret(deviceIdentification, t))
+                .map(s -> this.createEncrypted(deviceIdentification, s, this.getKey()))
+                .forEach(this.secretRepository::save);
+        //@formatter:on
+    }
+
+    public synchronized void activateNewSecrets(final String deviceIdentification, final List<SecretType> secretTypes) {
+        secretTypes.stream().forEach(t -> this.checkNrNewSecretsOfType(deviceIdentification, t, 1));
+        secretTypes.stream().forEach(t -> this.activateNewSecret(deviceIdentification, t));
+    }
+
+    public boolean hasNewSecret(final String deviceIdentification, final SecretType secretType) {
+        return this.secretRepository.getSecretCount(deviceIdentification, secretType, SecretStatus.NEW) > 0;
+    }
+
+    private synchronized long activateNewSecret(final String deviceIdentification, final SecretType secretType) {
+        final Optional<DbEncryptedSecret> activeSecretOptional = this.getSingleDbEncryptedSecret(deviceIdentification,
+                secretType, SecretStatus.ACTIVE);
+        if (activeSecretOptional.isPresent()) {
+            final DbEncryptedSecret currentSecret = activeSecretOptional.get();
+            currentSecret.setSecretStatus(SecretStatus.EXPIRED);
+            this.secretRepository.save(currentSecret);
+        }
+        final DbEncryptedSecret newSecret = this.getSingleDbEncryptedSecret(deviceIdentification, secretType,
+                SecretStatus.NEW).get(); //We have checked nr of new secrets already in activateNewSecrets
+        newSecret.setSecretStatus(SecretStatus.ACTIVE);
+        return this.secretRepository.save(newSecret).getId();
+    }
+
+    private void checkNrNewSecretsOfType(final String deviceIdentification, final SecretType t, final int expectedNr) {
+        final int nrNewSecretsOfType = this.secretRepository.getSecretCount(deviceIdentification, t, SecretStatus.NEW);
+        if (nrNewSecretsOfType != expectedNr) {
+            final String errorMsg = "Expected %s new secrets, but %s new secret(s) present";
+            throw new IllegalStateException(String.format(errorMsg, expectedNr, nrNewSecretsOfType));
+        }
+    }
+
+    public TypedSecret generateAes128BitsSecret(final SecretType secretType) {
+        final DbEncryptionKeyReference keyReference = this.getKey();
+        //TODO get key (KEK) from HSM and use it to encrypt generated AES 128-bits secret
+        //Old implementation:
+        try {
+            final KeyGenerator keyGenerator = KeyGenerator.getInstance("AES");
+            keyGenerator.init(AES_GMC_128_KEY_SIZE);
+            final byte[] encodedSecret = keyGenerator.generateKey().getEncoded();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this unused "encodedSecret" local variable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1481)